src/emqtt.erl

```diff
@@ -1215,6 +1316,85 @@ should_ping(ConnMod, Sock, _LastPktId) ->
             Error
     end.
 
+maybe_shoot(State0 = #state{pendings = Pendings}) ->
+    NPendings = drop_expired(Pendings),
+    State = State0#state{pendings = NPendings},
+    case {is_pendings_empty(NPendings), is_inflight_full(State)} of
+        {false, false}->
+            shoot(State#state{pendings = NPendings});
+        {true, _} ->
+            {keep_state, State};
+        {_, true} ->
+            {next_state, inflight_full, State}
+    end.
+
+shoot(State = #state{pendings = Pendings}) ->
+    {PubReq, NPendings} = dequeue_publish_req(Pendings),
+    shoot(PubReq, State#state{pendings = NPendings}).
+
+shoot(?PUB_REQ(Msg = #mqtt_msg{qos = ?QOS_0}, _ExpireAt, Callback), State) ->
+    case send(Msg, State) of
+        {ok, NState} ->
+            eval_callback_handler(ok, Callback),
+            maybe_shoot(NState);
+        {error, Reason} ->
+            shutdown_due_to_send_failed({error, Reason}, State)
+    end;
+
+shoot(?PUB_REQ(Msg = #mqtt_msg{qos = QoS}, ExpireAt, Callback),
+      State = #state{last_packet_id = PacketId, inflight = Inflight})
+  when QoS == ?QOS_1; QoS == ?QOS_2 ->
+    Msg1 = Msg#mqtt_msg{packet_id = PacketId},
+    case send(Msg1, State) of
+        {ok, NState} ->
+            Inflight1 = update_inflight(
+                          PacketId,
+                          ?PUB_REQ(Msg1, ExpireAt, Callback),
+                          Inflight
+                         ),
+            State1 = ensure_retry_timer(NState#state{inflight = Inflight1}),
+            maybe_shoot(bump_last_packet_id(State1));
+        {error, Reason} ->
+            shutdown_due_to_send_failed({error, Reason}, State)
+    end.
+
+shutdown_due_to_send_failed(
+  {error, Reason},
+  State = #state{pendings = Pendings, inflight = Inflight}) ->
+    %% reply to all pendings caller
+    Reqs = maps:get(requests, Pendings),
+    _ = queue:fold(
+          fun(?PUB_REQ(_, _, Callback), _) ->
+                  eval_callback_handler({error, Reason}, Callback)
+          end, ok, Reqs),
+    _ = maps:foreach(
+          fun(_PacketId, ?INFLIGHT_PUBLISH(_Msg, _SentAt, _ExpireAt, Callback)) ->
+                  eval_callback_handler({error, Reason}, Callback);
+             (_PacketId, _InflightPubReql) ->
+                  ok
+          end, Inflight),
```

Comment:
should reply inflight before queued.

Reply:
should reply inflight in a reversed order ?

Reply:
Ordered by packet id?  Then one would have to convert the map to an ordered list.

Reply:
I checked spec recently it didn't say the packet id must be incremental.

Reply:
It is true that there is no said, but the incremental I feel more comfortable to use? So I moved `bump_last_packet_id` from `send/x` into the`shoot/x` function🙂